### ISSUE TITLE
Connections to the OpenShift API are kept open and accumulating until no more connections can be established, turning Kiali inaccessible

### ISSUE THREAD

**Issue by hhovsepy** [kiali/kiali#2905]
Happens on Kiali latest master.
OCP 4.5
Istio 1.6.1

When logged in to Kiali from 2 different browsers (FF and Chrome), after the second browser login, the first browser session is timeout, and it sows either session timeout message, or "Error trying to get OAuth metadata".
Started happening recently and not constantly reproducible.
Only after some time login works for one of browsers.

![Screenshot from 2020-06-15 16-41-03](https://user-images.githubusercontent.com/604313/84671200-60276000-af27-11ea-841b-f7c28ed91d53.png)
![Screenshot from 2020-06-15 16-42-34](https://user-images.githubusercontent.com/604313/84671203-60bff680-af27-11ea-9a14-1b07e3c5d5c7.png)

[kiali-7d8d7dfcf9-rw89t-kiali.log](https://github.com/kiali/kiali/files/4781000/kiali-7d8d7dfcf9-rw89t-kiali.log)

**Comment by jmazzitelli** [comment]
These look to be the relevent logs - something odd with networking:

```
W0615 14:45:01.774090       1 authentication.go:523] Token error: could not get user info from Openshift: failed to get api endpoint apis/user.openshift.io/v1/users/~ for oauth consumption, error: Get "https://kubernetes.default.svc/apis/user.openshift.io/v1/users/~": dial tcp 172.30.0.1:443: connect: cannot assign requested address
E0615 14:45:01.774157       1 openshift_oauth.go:127] failed to get api endpoint apis/user.openshift.io/v1/users/~ for oauth consumption, error: Get "https://kubernetes.default.svc/apis/user.openshift.io/v1/users/~": dial tcp 172.30.0.1:443: connect: cannot assign requested address
W0615 14:45:01.774167       1 authentication.go:523] Token error: could not get user info from Openshift: failed to get api endpoint apis/user.openshift.io/v1/users/~ for oauth consumption, error: Get "https://kubernetes.default.svc/apis/user.openshift.io/v1/users/~": dial tcp 172.30.0.1:443: connect: cannot assign requested address
```

**Comment by israel-hdez** [comment]
So, the HTTP client is trying to open a connection but is failing. It's probably related to networking, but I'd lean towards a client with hanging connections up to the point where no more ports are available to successfully open a connection. 

@hhovsepy are you doing a particular test? or testing a particular page of Kiali?
May be there is a page in Kiali opening lots of requests to the cluster API up to the point that having two browsers is enough to reach the limit of connections between the Kiali pod and the cluster API (which should ge 65k connections)... Or, probably a request is hanging and leaving open connections leading to slowly reach the connection limit.

**Comment by hhovsepy** [comment]
There are 2 problems:
1. Your session is expired or was terminated in another window.
The first error happens when Kiali page is open and logged in in different browsers, but also Openshift console is also open on those browsers. Happens on OCP 4.5 and 4.4 as well.

2. Error trying to get OAuth metadata.
The Second issue happens only on OCP 4.5 and when UI automation testsuite is running in parallel with manually checking Kiali in browser.

Both issues are not constantly reproducible and started happening recently.

**Comment by israel-hdez** [comment]
Non-openshift cluster looks not affected.
Seems to be OpenShift-specific.

**Comment by israel-hdez** [comment]
Not OpenShift-specific. Also k8s is affected.

When querying metrics, connections to Prometheus hang for a while in TIME_WAIT state, anywhere between 1-4 minutes (depending on underlying system).

**Comment by israel-hdez** [comment]
When doing automated testing, things may happen quite quickly and hanging connections to Prometheus may pile up causing blockage of ANY outgoing connections (not only p8s) until those hanged connections get terminated.

**Comment by mattmahoneyrh** [comment]
Encountered this error while testing PR 2944 on  OCP 4.5.0-rc.6

[kiali-556b874d99-cpqt6-kiali.log](https://github.com/kiali/kiali/files/4892185/kiali-556b874d99-cpqt6-kiali.log)

**Comment by israel-hdez** [comment]
No luck fixing this. On hold until next sprint.

**Comment by jshaughn** [comment]
Has this been tried with caching off?

**Comment by israel-hdez** [comment]
I haven't!
I'm not sure it will make a difference, but worth to try, just in case.

**Comment by israel-hdez** [comment]
I'm releasing this issue because, AFAIK, it's been seen only in QE tests.
We, probably, can retake when it affects users.

**Comment by jshaughn** [comment]
Closing, can be re-opened if priority increases.

**Comment by israel-hdez** [comment]
Affecting users who are using OS 4.4.
Probably, OS 4.3 and prior versions are unaffected (I'm asking) as somebody started seeing this issue after upgrading to 4.4 from 4.3.

**Comment by israel-hdez** [comment]
Issue has been seen on OS 4.3, 4.4 and 4.5.
Reopening, because it's happening on non-QE (end-user) envs.

**Comment by jotak** [comment]
Did you setup a timeout on the http client connection? Perhaps it's totally unrelated but ... https://medium.com/@nate510/don-t-use-go-s-default-http-client-4804cb19f779

**Comment by jotak** [comment]
That might be the problem, I see in openshift_oauth.go there's no timeout configured, unlike in openid_auth.go.

**Comment by jotak** [comment]
PS: I see in versions.go / `istioVersion()` there's also a client connection that is done without timeout and can potentially lead to a leak. We should find a way to enforce using a timeout, if this is really the problem.

**Comment by jshaughn** [comment]
@jotak Thanks, we will try the suggestion.  We finally recreated the issue but didn't have any ideas about resolution.

**Comment by israel-hdez** [comment]
I'm changing title to denote that the issue was that connections between Kiali and the OpenShift API are the ones being dangling.

**Comment by bethmage** [comment]
Kiali operator  version 1.26.1 istio v1.7.5 on K8s 1.18.10
and oidc is Set-up with no RBAC support.

After successfull login with oidc
Your session is will expire in 60 seconds popup is shown after 180 seconds.
It is possible to login again and same happens !

Is some way to configure the session-timeout in kiali ? 
<img width="600" alt="Screenshot 2020-11-23 at 15 03 28" src="https://user-images.githubusercontent.com/2519342/99975752-eaab9f80-2da2-11eb-856b-6a19519b118e.png">

**Comment by israel-hdez** [comment]
The session timeout is given by the OpenId Server. You need to ask to your OIDC Server admin to extend the session timeout.

Out of curiosity, what OIDC server are you using?

**Comment by FilipB** [comment]
I see this error in the latest production SM:
```
Kiali UI
    1.24.1
Kiali Server
    v1.24.2.redhat0
Kiali Container
    1.24.2
OpenShift Service Mesh
    2.0.0 
```
I'm able to reproduce it using multiple bookinfo apps. With 16 bookinfo apps I can see the memory consumption of kiali pod to grow steadily until it reaches ~3GB. I can also see that number of open files (`lsof | grep kiali | wc -l`) on OCP node which is running the kiali pod is growing for the kiali process and containing hundreds of thousands of:
`kiali     3101962 3107794 kiali             10006100   18u     sock                0,9       0t0   20680725 protocol: TCP`
After some time the 
`E0708 17:07:33.139627       1 openshift_oauth.go:170] failed to get api endpoint apis/route.openshift.io/v1/namespaces/istio-system/routes/kiali for oauth consumption, error: Get "https://kubernetes.default.svc/apis/route.openshift.io/v1/namespaces/istio-system/routes/kiali": dial tcp 172.30.0.1:443: connect: cannot assign requested address`
error is thrown to the kiali log and auth is not working showing "Your session is expired or was terminated in another window."

When the kiali pod is restarted the issue is gone but the memory consumption and number of open files grow until it breaks again.

**Comment by FilipB** [comment]
Also `netstat  -plant | grep ESTABLISHED | wc -l` on the OCP node running kiali is only showing: 33 so it seems it's not caused by leaking TCP connections.

**Comment by bethmage** [comment]
@israel-hdez, Thank you.
I use Keycloak v 11.0.2 as an IDP.

I assumed it was the keycloak **SSO session idle** value in REALM and it is set to 30 Minutes.
For Authorization Code Flow, the value in **Access Token Lifespan** is relevant and was only 5 minutes.
Increasing the Access Token Lifespan in keycloak client Advance setting solved the problem.

sorry, this blog may not be the right place for such problems

**Comment by israel-hdez** [comment]
@bethmage No problem. Happy to know it's fixed.

**Comment by israel-hdez** [comment]
@FilipB I'm checking the git history, and this fix landed upstream in v1.24.3.
I remember @jmazzitelli sending the notification of the builds, but AFAIK it's only unsigned. Perhaps, mazz can comment if 1.24.3 will be fully ready soon.

**Comment by FilipB** [comment]
@israel-hdez Thank you Edgar. I tried current master and it's working fine. Memory is not growing. I will also try v1.24.3 and file downstream issue if it's still visible there.
